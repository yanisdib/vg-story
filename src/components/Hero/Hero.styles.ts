import styled from 'styled-components';


interface StyledSectionProps {
    readonly background?: string;
}


export const Section = styled.section<StyledSectionProps>`
    background-color: ${({
        background
    }: StyledSectionProps) => background};

    @media (min-width: 375px) {
        display: flex;
        flex-direction: row;
        height: 100vh;

    }

    @media (min-width: 950px) {
        height: 95vh;
    }
`;

export const Wrapper = styled.div`
    flex-direction: column;
    width: 100%;
    height: inherit;
`;

export const Content = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;
    width: inherit;
    height: inherit;
`;

export const Preview = styled.div`
    display: flex;
    flex-direction: column;
    flex-basis: 50%;
    justify-content: center;
    align-items: center;
    height: inherit;
    padding: 0 calc(var(--ui-padding-y) - 25px);
    
    a {
        color: #000;
        text-decoration: none;

        &:hover {
            text-decoration: underline;
        }
    }

    h1 {
        margin-bottom: 22px;
    }
    
    h4 {
        align-self: flex-start;
        margin: 10px 0;

        span > a {
            text-decoration: underline;

            &:hover{
                text-decoration: none;
            }
        }
    }
`;

export const FrontImage = styled.div`
    display: flex;
    flex-direction: column;
    /* flex-basis: 50%; */
    justify-content: center;
    align-items: center;
    height: inherit;
    background-color: #FFF;

    a {
        height: 100%;

        img {
            width: 100%;
            height: 100%;
            object-fit: cover;
            object-position: center;
        }
    }
`;
