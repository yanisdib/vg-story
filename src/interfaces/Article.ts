import {
    Comment,
    ImageAttributes,
    Topic
} from '.';


/**
 * Article is an interface which describes properties
 * an article should be created with
 */
export interface Article {
    author: string;
    backgroundHex?: string;
    body?: string;
    comments?: Comment[];
    createdAt: number;
    editedAt?: number;
    frontImage: ImageAttributes;
    imageGallery?: {
        id: string,
        images: ImageAttributes[]
    }[];
    id: string;
    isFeatured?: boolean;
    permalink: string;
    title: string;
    topics: Topic[];
}