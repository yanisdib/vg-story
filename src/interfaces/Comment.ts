/**
 * Comment is an interface which describes properties
 * an article's comment should be created with
 */
export interface Comment {
    id: string;
    body: string;
    createdAt: number;
}