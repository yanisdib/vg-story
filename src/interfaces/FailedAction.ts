import { Action } from '@reduxjs/toolkit';


/**
 * FailedAction is an interface which describes
 * properties returned by the payload of
 * an action that failed
 */
export interface FailedAction extends Action {
    error: Error;
}