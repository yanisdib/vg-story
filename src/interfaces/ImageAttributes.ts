/**
 * Attributes an image object should be declared with
 */
export interface ImageAttributes {
    alt?: string;
    src: string;
    title?: string;
}