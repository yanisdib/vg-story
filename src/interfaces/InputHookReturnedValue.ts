export type InputHookReturnedValue<T> = [
    T,
    (event: React.ChangeEvent<HTMLInputElement>) => void
]
