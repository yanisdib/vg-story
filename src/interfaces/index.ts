export type { Article } from './Article';
export type { Comment } from './Comment';
export type { ImageAttributes } from './ImageAttributes';
export type { Topic } from './Topic';

export type { FailedAction } from './FailedAction';
export type { StoreState } from './StoreState';
