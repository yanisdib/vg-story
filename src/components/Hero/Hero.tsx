import { Link } from 'react-router-dom';

import { Article } from '../../interfaces';

import {
    Content,
    FrontImage,
    Preview,
    Section,
    Wrapper
} from './Hero.styles';


interface HeroProps {
    readonly article: Article;
}


function Hero({ article }: HeroProps) {
    
    const {
        backgroundHex = '#000',
        frontImage,
        permalink,
        title,
        topics
    } = article;

    return (
        <Section
            data-testid="hero-section"
            background={backgroundHex}
        >
            <Wrapper>
                <Content>
                    <Preview>
                        <h4>
                            {topics
                                .map(topic => topic.name)
                                .join(' â€” ')
                            }
                        </h4>
                        <h1>
                            <Link to={`article/${permalink}`}>
                                {title}
                            </Link>
                        </h1>
                        <h4>
                            <span>
                                <Link to={`article/${permalink}`}>
                                    Read -&gt;
                                </Link>
                            </span>
                        </h4>
                    </Preview>
                    <FrontImage>
                        <Link to={`article/${permalink}`} title="Read">
                            <img
                                alt={`${frontImage?.alt ?? `Article's image`}`}
                                src={frontImage?.src}
                                title={`${frontImage?.title ?? `Article's image`}`}
                            />
                        </Link>
                    </FrontImage>
                </Content>
            </Wrapper>
        </Section>
    );
}


export default Hero;