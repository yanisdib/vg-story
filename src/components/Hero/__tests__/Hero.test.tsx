import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom';

import { Article } from '../../../interfaces';

import Hero from '../Hero';


describe('Hero component', () => {

    const article: Article = {
        author: 'Yanis',
        backgroundHex: '#FF5B5B',
        frontImage: { src: 'hello.jpg' },
        createdAt: 1000,
        id: '12de',
        permalink: 'this-is-a-test',
        title: 'This is a test',
        topics: [
            {
                createdAt: 0,
                id: '1ac',
                name: 'Test',
                permalink: 'test'
            }
        ]
    };

    it('should render without crashing', () => {

        const { container } = render(
            <Hero article={article} />,
            { wrapper: MemoryRouter }
        );

        expect(container).toMatchSnapshot();
    });

    it('should render title of article', () => {

        const { getByText } = render(
            <Hero article={article} />,
            { wrapper: MemoryRouter }
        );

        expect(getByText('This is a test')).toBeInTheDocument();
    });

    it('should have background color of the article', () => {

        const { getByTestId } = render(
            <Hero article={article} />,
            { wrapper: MemoryRouter }
        );

        expect(getByTestId('hero-section'))
            .toHaveStyle(`background-color: ${article.backgroundHex}`);
    });

    it('should have a black background by default', () => {
        const testArticle: Article = {
            author: 'Yanis',
            frontImage: { src: 'hello.jpg' },
            createdAt: 1000,
            id: '12de',
            permalink: 'this-is-a-test',
            title: 'This is a test',
            topics: [
                {
                    createdAt: 0,
                    id: '1ac',
                    name: 'Test',
                    permalink: 'test'
                }
            ]
        };

        const { getByTestId } = render(
            <Hero article={testArticle} />,
            { wrapper: MemoryRouter }
        );

        expect(getByTestId('hero-section')).toHaveStyle(`background-color: #000`)
    });
});