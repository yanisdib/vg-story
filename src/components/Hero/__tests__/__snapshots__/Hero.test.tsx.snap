// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero component should render without crashing 1`] = `
<div>
  <section
    class="sc-bczRLJ qjPlD"
    data-testid="hero-section"
  >
    <div
      class="sc-gsnTZi dqCHuM"
    >
      <div
        class="sc-dkzDqf kdZSXg"
      >
        <div
          class="sc-hKMtZM fVTBvs"
        >
          <h4>
            Test
          </h4>
          <h1>
            <a
              href="/article/this-is-a-test"
            >
              This is a test
            </a>
          </h1>
          <h4>
            <span>
              <a
                href="/article/this-is-a-test"
              >
                Read -&gt;
              </a>
            </span>
          </h4>
        </div>
        <div
          class="sc-eCYdqJ eOJYOC"
        >
          <a
            href="/article/this-is-a-test"
            title="Read"
          >
            <img
              alt="Article's image"
              src="hello.jpg"
              title="Article's image"
            />
          </a>
        </div>
      </div>
    </div>
  </section>
</div>
`;
